### PR TITLE
Enhancement webots js textures

### PR DESCRIPTION
**Pull request by stefaniapedrazzi**
Addresses #749:

Improve scene load and rendering performance:
* [x] cache and reuse Texture objects when possible, i.e. when the texture transform, wrap etc. matches.
* [x] if textures are down scaled due to the texture quality settings then export the down scaled images to the web interface instead of the original ones
* [x] fix navigation in the scene not smooth on scene load -> disable frustum culling until all the textures are loaded
* [x] improve animation: automatically start it on page load only when all the textures are loaded to avoid slowing down the animation due to the texture computations and renderings.

~~Bug fix:~~
* [x] ~~fix issue due to GpuPicker not supporting undefined `scene.background` (change `undefined` to `null`)~~ _edit: fixed in #766_

### CODE REVIEW

**Comment by stefaniapedrazzi** [comment]
In the `obstacle_avoidance` benchmark creating a single `THREE.Texture` instance, if the texture parameters matches, reduces the instances for the `VarnishedPine` PBRAppearance (4 textures with resolution 2048x2048) from 30 to just 5.

**Comment by stefaniapedrazzi** [comment]
The rendering clearly takes long when new objects enter the viewpoint frustum (for example the windows and the professor table in the obstacle_avoidance benchmark).
If two objects share the same textures (for example the windows on the left and right wall), only the first object will slow down the scene navigation. Given that just the textures and not the geometries are reused it seems that the problem are again textures and materials setup.

I tried different solutions on scene load to try to  avoid this slow down when navigating into the scene:
* set the `Object3D.frustumCulling` of each object to true
* call `WebGLRenderer.compile()`
but on my system I don't see any difference.

**Comment by omichel** [comment]
That's unfortunately another drawback of using Threejs versus using Wren.js. Resolving this problem in Wren.js would have been easier (e.g., similar to what we did in Webots to fix the problem). However, there should be a way in Threejs to force load the textures. On Webots/Wren, I recall we disabled the culling a forced one rendering, which caused Wren to load all the textures. Can't we do something similar in Threejs?

**Comment by stefaniapedrazzi** [comment]
> However, there should be a way in Threejs to force load the textures. On Webots/Wren, I recall we disabled the culling a forced one rendering, which caused Wren to load all the textures. Can't we do something similar in Threejs?

This is what the `frustumCulling = true` should be supposed to do. But it doesn't solve the issue.

**Comment by omichel** [comment]
That sounds odd, I would rather disable the frustum culling instead of enabling it to ensure all the objects are rendered.

**Comment by stefaniapedrazzi** [comment]
> That sounds odd, I would rather disable the frustum culling instead of enabling it to ensure all the objects are rendered.

This is the way to disable/enable frustum culling on three.js:
https://github.com/mrdoob/three.js/issues/3949

**Comment by omichel** [comment]
Yes, it should be set to `false`, not `true`, isn't it?

**Comment by stefaniapedrazzi** [comment]
Yes. I wrote the wrong value in the doc.
But I tested with `false` and it doesn't solve the issue.

**Comment by stefaniapedrazzi** [comment]
oops finally I had a typo in the code. That's working.

**Comment by stefaniapedrazzi** [comment]
:+1:

**Comment by fabienrohrer** [comment]
👀

File: resources/web/wwi/texture_loader.js [hunk 1]
```diff
@@ -184,6 +265,43 @@ class _TextureLoaderObject {
...
       this.onTextureLoad();
+
+    this._evaluatePendingData();
+  }
+
+  _setTimeout() {
+    // Set texture loading timeout.
+    // If after some time no new textures are loaded, the hasPendingData variable is automatically
+    // reset to false in order to handle not found textures.
+    // The `this.loadingTextures` dictionary is not reset so that it is still possible to recore late textures.
```
Comment (omichel) [inline resources/web/wwi/texture_loader.js]: What does "recore" mean? I guess it's a typo.

Comment (stefaniapedrazzi) [reply resources/web/wwi/texture_loader.js]: yes. I don't remember what I meant, but `load` is fine.

**Pull request merged**